kube-client/src/client/tls.rs

```diff
@@ -53,109 +53,93 @@ pub mod native_tls {
-                        })
-                    })
-                    .map_err(|_| Error::SslError("No valid private key was found".into()))?;
-
-                if let (Some(sk), false) = (sk.pop(), certs.is_empty()) {
-                    (sk, certs)
-                } else {
-                    return Err(Error::SslError("private key or certificate not found".into()));
-                }
-            };
+        let config_builder = ClientConfig::builder()
+            .with_safe_defaults()
+            .with_root_certificates(root_store(root_certs)?);
+
+        let mut client_config = if let Some((chain, pkey)) = identity_pem.map(client_auth).transpose()? {
+            config_builder
+                .with_single_cert(chain, pkey)
+                .map_err(|e| Error::SslError(format!("{}", e)))?
+        } else {
+            config_builder.with_no_client_auth()
+        };
 
+        if accept_invalid {
             client_config
-                .set_single_client_cert(certs, key)
-                .map_err(|e| Error::SslError(format!("{}", e)))?;
+                .dangerous()
+                .set_certificate_verifier(std::sync::Arc::new(NoCertificateVerification {}));
         }
+        Ok(client_config)
+    }
 
-        if let Some(ders) = root_cert {
+    fn root_store(root_certs: Option<&[Vec<u8>]>) -> Result<rustls::RootCertStore> {
+        let mut root_store = rustls::RootCertStore::empty();
+        if let Some(ders) = root_certs {
             for der in ders {
-                client_config
-                    .root_store
-                    .add(&Certificate(der.to_owned()))
+                root_store
+                    .add(&Certificate(der.clone()))
                     .map_err(|e| Error::SslError(format!("{}", e)))?;
             }
         }
+        Ok(root_store)
+    }
 
-        if accept_invalid {
-            client_config
-                .dangerous()
-                .set_certificate_verifier(Arc::new(NoCertificateVerification {}));
+    // TODO Support EC Private Key to support k3d. Need to convert to PKCS#8 or RSA (PKCS#1).
+    // `openssl pkcs8 -topk8 -nocrypt -in ec.pem -out pkcs8.pem`
+    // https://wiki.openssl.org/index.php/Command_Line_Elliptic_Curve_Operations#EC_Private_Key_File_Formats
+    fn client_auth(data: &[u8]) -> Result<(Vec<Certificate>, PrivateKey)> {
+        use rustls_pemfile::Item;
+
+        let mut cert_chain = Vec::new();
+        let mut pkcs8_key = None;
+        let mut rsa_key = None;
+        let mut reader = std::io::Cursor::new(data);
+        for item in rustls_pemfile::read_all(&mut reader)
+            .map_err(|e| Error::SslError(format!("failed to read identity PEM: {}", e)))?
+        {
+            match item {
+                Item::X509Certificate(cert) => cert_chain.push(Certificate(cert)),
+                Item::PKCS8Key(key) => pkcs8_key = Some(PrivateKey(key)),
+                Item::RSAKey(key) => rsa_key = Some(PrivateKey(key)),
+            }
```

Comment:
Changed to get all sections at once, but it should behave the same. Use the last found private key, and PKCS8 key is preferred.